---
title: "ACS demo"
output: html_document
---

Installation and practical usage of the acs R package.

Install & Load the library
```{r}
install.packages("acs")
library(acs)
```

Get an API key from <http://api.census.gov/data/key_signup.html>
Mine is b9cfb06fa75169602cd01ece4bef67ae71654b93

Set the API key
```{r}
acs::api.key.install(key='b9cfb06fa75169602cd01ece4bef67ae71654b93')
# A blank API key is also known to work.
# acs::api.key.install(key='')
```

Set up the zip-code based geography
```{r}

# Get zip codes for CA
library(visdom)
data(package="visdom", list="ACS_11_5YR_DP04")
zip_codes = ACS_11_5YR_DP04$ZCTA

# This calling style of limiting results to a state works with census tracts, but not zip codes. At least not for the ACS datasets.
# geography = geo.make(state='CA', zip.code='*', combine=F)

# This works, but pulls down all zip codes for the US
# geography = geo.make(zip.code='*', combine=F)

# This fails, apparently zip.code doesn't like taking a list.
# geography = geo.make(zip.code=zip_codes, combine=F)

# This works
geography = c(geo.make(zip.code=96161), geo.make(zip.code=97635))

# This doesn't work with acs.fetch because geography is a list instead of a geo.set
# geo.make.zip = function(zip) return(geo.make(zip.code=zip))
# geography = sapply(X=zip_codes, FUN=geo.make.zip)

# This kind of works as long as I initialize geography as a geo.set
geography = geo.make(zip.code=zip_codes[1]))
for (zip in zip_codes[2:length(zip_codes)]) {
     geography = c(geography, geo.make(zip.code=zip))
}

# It crashed out while trying to iterate through all of the CA zip codes at once.
# Doing API calls with smaller chunks works.
geography = geography[1:10]
```

Fetch data
```{r}

# Use the acs canonical example of table B01003
results = acs.fetch(
     endyear=2011, span = 5,
     geography=geography,
     table.number="B01003", # The DP tables don't work... table.number='DP02', 
     dataset = "acs")
estimate(results)
standard.error(results)

# Same results, but ask for the variable directly
results = acs.fetch(
     endyear=2011,
     span = 5,
     geography=geography,
     variable=c('B01003_001'), 
     dataset = "acs")
estimate(results)
standard.error(results)

# Now try with a variable that we care about from the Data Profile tables
names(ACS_11_5YR_DP04)[4] # "HC01_VC03"
results = acs.fetch(
     endyear=2011,
     span = 5,
     geography=geography, 
     variable=c('HC01_VC03'), 
     dataset = "acs")
# Failure!
# Error in file(file, "rt") : cannot open the connection
# In addition: Warning message:
# No data found at:
#   http://api.census.gov/data/2011/acs5?key=b9cfb06fa75169602cd01ece4bef67ae71654b93&get=HC01_VC03E,HC01_VC03M,NAME&for=zip+code+tabulation+area:89010 

```

It turns out the acs package cannot access the 'Data Profile' tables which constitute all of the variables that we care about. From the census ACS API website <http://www.census.gov/data/developers/data-sets/acs-5year.html>, I see that the "Summary File" and "Data Profile" are accessed differently. 
http://api.census.gov/data/2014/acs5?...
vs
http://api.census.gov/data/2014/acs5/profile?...

I also notice that "Data Profile" data is only available from 2012 onwards, and that the names of the Data Profile variables changed from 2011 to 2012. 

After consulting with Sam, I proceeded by updating the acs package to work with the Data Profile tables if you ask for the acs_dp dataset. This is available on my personal github account and is submitted as a pull request to the package maintainers. 